### PR TITLE
Fix code quality issues in sof tools

### PR DESCRIPTION
**Pull request by softwarecki**
Fixed many minor bugs in sof tools applications.

### CODE REVIEW

**Review by kv2019i — APPROVED** [review]
Thank you @softwarecki , looks good to go!

**Review by jsarha — APPROVED** [review]
Could not find anything else wrong but this commit message appears to be somehow truncated:

logger: convert: Fixed handling of an error reported by clock_gettime
The clock_gettime function only returns information that an error occurred.
The error code should be taken from the e

Signed-off-by: Adrian Warecki <adrian.warecki@intel.com>

**Comment by marc-hb** [comment]
Please submit this to stable-v2.2 for actual testing

File: tools/logger/convert.c [hunk 1]
```diff
@@ -349,16 +347,14 @@ static inline void print_table_header(void)
...
 
 	if (global_config->time_precision >= 0) {
-		const unsigned int ts_width =
-			timestamp_width(global_config->time_precision);
-		snprintf(time_fmt, sizeof(time_fmt), "%%-%ds(us)%%%ds  ",
-			 ts_width, ts_width);
-		fprintf(out_fd, time_fmt, " TIMESTAMP", "DELTA");
+		const unsigned int ts_width = timestamp_width(global_config->time_precision);
+
+		fprintf(out_fd, "%*s(us)%*s  ", -ts_width, " TIMESTAMP", ts_width, "DELTA");
```
Comment (marc-hb) [inline tools/logger/convert.c:357]: No, this is not equivalent at all. We already discussed this:
- https://github.com/thesofproject/sof/pull/6738#discussion_r1042797977

This PR has clearly not been tested.

**Review by marc-hb — CHANGES_REQUESTED** [review]

**Comment by marc-hb** [comment]
> Please submit this to stable-v2.2 for actual testing

BTW smex is not part of the binary release, has never been. It's only part of the build process.

The incoming 2023.9, Intel's sof-bin release will include the sof-logger from... the 2.2 branch. So sof-logger fixes in the main branch will not affect it.

File: smex/elf.c [hunk 2]
```diff
@@ -408,10 +408,17 @@ int elf_find_section(const struct elf_module *module, const char *name)
 		ret = -errno;
 		goto out;
 	}
+	buffer[section->size - 1] = '\0';
```
Comment (btian1) [inline smex/elf.c:411]: check commit message, seems "doesn‘t not" have something wrong?

File: tools/logger/convert.c [hunk 3]
```diff
@@ -349,7 +349,7 @@ static inline void print_table_header(void)
 
 	if (gettime_ret) {
 		log_err("clock_gettime() failed: %s\n",
-			strerror(gettime_ret));
+			strerror(errno));
```
Comment (btian1) [inline tools/logger/convert.c:350]: still the commit message, "from the errorno", it is a global variables..

File: tools/logger/convert.c [hunk 4]
```diff
@@ -624,7 +624,13 @@ static int read_entry_from_ldc_file(struct ldc_entry *entry, uint32_t log_entry_
 	entry->params = NULL;
 
 	/* set file position to beginning of processed entry */
-	fseek(global_config->ldc_fd, entry_offset, SEEK_SET);
+	ret = fseek(global_config->ldc_fd, entry_offset, SEEK_SET);
+	if (ret) {
+		log_err("Failed to seek to entry header for offset 0x%x in dictionary.\n",
+			entry_offset);
+		ret = -1;
```
Comment (btian1) [inline tools/logger/convert.c:621]: line 631 does not need, ret either 0 or -1.

File: tools/logger/convert.c [hunk 5]
```diff
@@ -915,8 +921,13 @@ static int logger_read(void)
 			/* When the address is not correct, move forward by one DWORD (not
 			 * entire struct dma_log)
 			 */
-			fseek(global_config->in_fd, -(sizeof(dma_log) - sizeof(uint32_t)),
+			ret = fseek(global_config->in_fd, -(sizeof(dma_log) - sizeof(uint32_t)),
 			      SEEK_CUR);
+			if (ret) {
+				log_err("fetch_entry() failed on seek, aborting\n");
+				ret = -errno;
```
Comment (btian1) [inline tools/logger/convert.c:921]: ret already assigned as -1

File: tools/logger/convert.c [hunk 6]
```diff
@@ -1101,7 +1112,12 @@ int convert(void)
 	}
 
 	/* read uuid section header */
-	fseek(config->ldc_fd, logs_hdr->data_offset + logs_hdr->data_length, SEEK_SET);
+	ret = fseek(config->ldc_fd, logs_hdr->data_offset + logs_hdr->data_length, SEEK_SET);
+	if (ret) {
+		log_err("Error while seeking to uuids header from %s.\n", config->ldc_file);
+		return -errno;
```
Comment (btian1) [inline tools/logger/convert.c:1111]: simply return ret;

File: tools/logger/logger.c [hunk 7]
```diff
@@ -220,17 +220,15 @@ static void *wait_open(const char *watched_dir, const char *expected_file)
 	const int dwatch = inotify_add_watch(iqueue, watched_dir, IN_CREATE);
 	struct stat expected_stat;
 	void *ret_stream = NULL;
-
-	char * const fpath = malloc(strlen(watched_dir) + 1 + strlen(expected_file) + 1);
+	const int fpath_len = strlen(watched_dir) + 1 + strlen(expected_file) + 1;
```
Comment (btian1) [inline tools/logger/logger.c:233]: need a static prefix?

File: src/audio/module_adapter/iadk/module_initial_settings_concrete.cpp [hunk 8]
```diff
@@ -49,6 +49,9 @@ ModuleInitialSettingsConcrete::ModuleInitialSettingsConcrete(DwordArray const &c
 		/* It shall contain BaseModuleCfg + BaseModuleCfgExt +       */
 		/* optionally some InputPinFormat[] + OutputPinFormat[] data */
 		CompoundCfg const * unvalidated_compound_cfg = cfg_ipc_msg.dataAs<CompoundCfg>();
+		if (!unvalidated_compound_cfg)
+			return;
```
Comment (btian1) [inline src/audio/module_adapter/iadk/module_initial_settings_concrete.cpp:53]: this was submitted as a separate pr: #8223

File: tools/logger/convert.c (hunk 1 above)
Comment (softwarecki) [reply tools/logger/convert.c:357]: In fact, it was done wrong in the linked PR. Run the code below and you will see that both approaches print the same result:

```
char time_fmt[255];
const unsigned int ts_width = 20;
    
snprintf(time_fmt, sizeof(time_fmt), "%%-%ds(us)%%%ds  ", ts_width, ts_width);
printf(time_fmt, " TIMESTAMP", "DELTA");
printf("\n");
printf("%*s(us)%*s  ", -ts_width, " TIMESTAMP", ts_width, "DELTA");
printf("\n");
```

File: tools/logger/convert.c (hunk 4 above)
Comment (softwarecki) [reply tools/logger/convert.c:621]: fread function returns
> Number of objects read successfully
so in case on an error ret will be 0, not -1.

File: tools/logger/convert.c (hunk 5 above)
Comment (softwarecki) [reply tools/logger/convert.c:921]: i want return error code set by fseek in errno variable.

File: tools/logger/logger.c (hunk 7 above)
Comment (softwarecki) [reply tools/logger/logger.c:233]: Why fpath_len variable should be a static?

**Comment by kv2019i** [comment]
@marc-hb @btian1 please re-review. We know this is not part of the firmware code, but improving code quality (both actual like resource leakage on error paths, plus more nice-to-have cases where we have code that is correct but irks pretty much all static code analysis tools) is always welcome.

**Comment by marc-hb** [comment]
> We know this is not part of the firmware code, but improving code quality [...] is always welcome.

Absolutely. However code quality implies running the code changes at least once in at least the default configuration.

>  (both actual like resource leakage on error paths, plus more nice-to-have cases where we have code that is correct but irks pretty much all static code analysis tools) 

I'm totally on board with making code simpler not just for humans but also for analyzers - even when it's already correct. Both usually go together. However there will always be some tricky corner cases and false positives and all analyzers offer a process to dispose those. Dynamically constructing a format string is likely one of these cases.

**Comment by kv2019i** [comment]
Backport of 7/10 test and submission to stable-v2.2 where we have more coverage in Intel CI for these tools https://github.com/thesofproject/sof/pull/8231

File: tools/logger/convert.c (hunk 1 above)
Comment (marc-hb) [reply tools/logger/convert.c:357]: Sincere apologies, I had been "burned" by #6738 and this comment was rushed. I didn't know about `%*s` and it looks great.

**Comment by marc-hb** [comment]
> Backport of 7/10 test and submission to stable-v2.2 where we have more coverage in Intel CI for these tools https://github.com/thesofproject/sof/pull/8231

All 10/10 commits here backported to stable-v2.2 just for testing in #8234

File: tools/logger/convert.c (hunk 1 above)
Comment (softwarecki) [reply tools/logger/convert.c:357]: No problem, that's what I thought you did it automatically. We discover new things all our lives :)

**Comment by kv2019i** [comment]
Thanks all, the tests with backported 2.2 look good ( https://github.com/thesofproject/sof/pull/8234 ). So only thing missing is a clean run of the mandatory CI. @softwarecki , can you push a rebase and/or ping CI folks to rerun -- the fail is very unlikely to be related to be caused by this PR, but we still need a clean run.

**Comment by marc-hb** [comment]
Two windows-build failed with "ninja not found" in https://github.com/thesofproject/sof/actions/runs/6248871611/job/16964356307?pr=8216

Doesn't really make sense because other Windows builds ran just fine.

Hopefully just a Github glitch, there was some https://www.githubstatus.com/history incident yesterday. Re-running the failed jobs.

**Review by btian1 — APPROVED** [review]

File: smex/elf.c [hunk 9]
```diff
@@ -451,17 +451,25 @@ int elf_read_section(const struct elf_module *module, const char *section_name,
...
+	if (ret) {
+		fprintf(stderr, "error: can't seek to %s section %d\n", section_name, -errno);
+		ret = -errno;
+		goto error;
+	}
+
+	ret = fread(*dst_buff, 1, section->size, module->fd);
+	if (ret != section->size) {
+		fprintf(stderr, "error: can't read %s section %d\n", section_name, -errno);
+		ret = -ENODATA;
```
Comment (lyakh) [inline smex/elf.c:471]: I'd do something like `ret = ret < 0 ? -errno : -ENODATA;` but at least this doesn't miss errors

**Review by lgirdwood — APPROVED** [review]

**Comment by softwarecki** [comment]
> @softwarecki can you check CI results. Thanks

I fixed one checkpatch issue.

Its look like we have some problem with CI for windows platform...

**Comment by lgirdwood** [comment]
> > @softwarecki can you check CI results. Thanks
> 
> I fixed one checkpatch issue.
> 
> Its look like we have some problem with CI for windows platform...

Yep, I'm surprised I'm not hearing any Windows developer complain here unless the Windows build config locally has Ninja and is out of sync with the container ?

**Comment by marc-hb** [comment]
> Yep, I'm surprised I'm not hearing any Windows developer complain here unless the Windows build config locally has Ninja and is out of sync with the container ?

It looks like an intermittent Github glitch, see above. Try re-running that one.

**Comment by marc-hb** [comment]
> Its look like we have some problem with CI for windows platform...

Filed new
- #8250

**Comment by kv2019i** [comment]
@softwarecki We need mandatory CIs to pass, this may need a repush (or manually trigger at quickbuild).

**Comment by softwarecki** [comment]
@kv2019i CI is all green now :) (except checkpatch which complains about camel case in Elf32_Shdr)

**Comment by lgirdwood** [comment]
> @softwarecki We need mandatory CIs to pass, this may need a repush (or manually trigger at quickbuild).

@wszypelt 

> @kv2019i CI is all green now :) (except checkpatch which complains about camel case in Elf32_Shdr)

@softwarecki can you check internal CI. I think we are good once it's green.
